utils/s3/client.cc

```diff
@@ -234,26 +235,84 @@ future<> map_s3_client_exception(std::exception_ptr ex) {
         auto e = std::current_exception();
         return make_exception_future<>(storage_io_error(EIO, format("S3 error ({})", e)));
     }
+}
+
+future<aws::aws_error> client::look_for_error(const http::reply& reply, input_stream<char>& in_, multipart_upload_completion is_mpu_completion) {
+    if (reply._status >= http::reply::status_type::multiple_choices || is_mpu_completion == multipart_upload_completion::yes) {
+        auto response_body = co_await util::read_entire_stream_contiguous(in_);
 
+        std::optional<aws::aws_error> possible_error = aws::aws_error::parse(std::move(response_body));
+        if (possible_error) {
+            co_return possible_error.value();
+        }
+        if (reply._status >= http::reply::status_type::multiple_choices) {
+            co_return aws::aws_error::from_http_code(reply._status);
+        }
+    }
+
+    co_return aws::aws_error{aws::aws_error_type::OK, aws::retryable::no};
 }
 
-future<> client::make_request(http::request req, http::experimental::client::reply_handler handle, http::reply::status_type expected) {
+future<> client::do_retryable_request(group_client& gc, http::request req, http::experimental::client::reply_handler handler) const {
+    uint32_t retries = 0;
+    do {
+        std::exception_ptr e;
+        aws::aws_exception request_ex{aws::aws_error{}};
+        try {
+            co_return co_await gc.http.make_request(req, handler, std::nullopt);
+        } catch (const aws::aws_exception& ex) {
+            ++retries;
+            e = std::current_exception();
+            request_ex = ex;
+        }
+        if (_retry_strategy->should_retry(request_ex.error(), retries)) {
+            s3l.debug("S3 client request failed. Reason: {}. Retry# {}", request_ex.what(), retries);
+            co_await seastar::sleep(_retry_strategy->delay_before_retry(request_ex.error(), retries));
+        } else {
+            if (!request_ex.error().is_retryable()) {
+                s3l.error("S3 client encountered non-retryable error. Reason: {}. Code: {}. Retry# {}", request_ex.what(),
+                        std::to_underlying(request_ex.error().get_error_type()), retries);
+            } else {
+                s3l.error("Retries exhausted. Retry# {}", retries);
+            }
+            co_await map_s3_client_exception(e);
```

Comment:
It's not obvious that this line actually makes this function resolve exceptional future. It _was_ when it was used inside .handle_exception() continuations, now it takes time to understand that once called it prevents the enclosing do{}while(true) from continuing

Reply:
hope it is clearer now